python/lsst/meas/extensions/ngmix/emPsfApprox.py

```diff
@@ -0,0 +1,152 @@
+    ]
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.SHAPE_ORDER
+
+    def __init__(self, config, name, schema, metadata):
+        SingleFramePlugin.__init__(self, config, name, schema, metadata)
+
+        self.flagHandler = FlagHandler.addFields(schema, name, FlagDefinitionVector(self.FLAGDEFS))
+        self.iterKey = schema.addField(name + '_iterations', type=int, doc="number of iterations run")
+        self.fdiffKey = schema.addField(name + '_fdiff', type=float, doc="fit difference")
+
+        # Add ShapeletFunction keys for the number of Gaussians requested
+        self.keys = []
+        for i in range(config.nGauss):
+            key = lsst.shapelet.ShapeletFunctionKey.addFields(schema,
+                  "%s_%d"%(name, i), "ngmix EM gaussian", "pixels", "",
+                  0, lsst.shapelet.HERMITE)
+            self.keys.append(key)
+        self.msfKey = lsst.shapelet.MultiShapeletFunctionKey(self.keys)
+
+    def measure(self, measRecord, exposure):
+
+        if exposure.getPsf() is None:
+            raise MeasurementError(self.flagHandler.getDefinition(self.NO_PSF).doc, self.NO_PSF)
+        psfArray = exposure.getPsf().computeKernelImage().getArray()
+        psfObs = ngmix.observation.Observation(psfArray,
+            jacobian=ngmix.UnitJacobian(row=(psfArray.shape[0]-1)/2, col=(psfArray.shape[1]-1)/2))
+
+        #  Need a guess at the sum of the diagonal moments
+        nGauss = self.config.nGauss
+        shape = exposure.getPsf().computeShape()
+        Tguess = shape.getIxx() + shape.getIyy()
+        emPars={'maxiter':self.config.maxIter, 'tol':self.config.tolerance}
+
+        runner=EMRunner(psfObs, Tguess, nGauss, emPars)
+        runner.go(ntry=self.config.nTries)
+        fitter=runner.get_fitter()
+        res=fitter.get_result()
+
+        #   Set the results, including the fit info returned by EMRunner
+        measRecord.set(self.iterKey, res['numiter'])
+        measRecord.set(self.fdiffKey, res['fdiff'])
+
+        #   We only know about two EM errors.  Anything else is thrown as "unknown".
+        if res['flags'] != 0:
+            if res['flags'] & 1:
+                raise MeasurementError(self.flagHandler.getDefinition(self.EM_RANGE_ERROR).doc, self.EM_RANGE_ERROR)
+            if res['flags'] & 2:
+                raise MeasurementError(self.flagHandler.getDefinition(self.EM_MAXITER).doc, self.EM_MAXITER)
+            raise RuntimeError("Unknown EM fitter exception")
+
+        #   Convert the nGauss Gaussians to ShapeletFunction's.  Zeroth order HERMITES are Gaussians.
+        #   There are always 6 parameters for each Gaussian.
+        psf_pars = fitter.get_gmix().get_full_pars()
+        for i in range(self.config.nGauss):
+            pars = psf_pars[i*6:i*6+6]
+            flux = pars[0]
+            x = pars[1]
+            y = pars[2]
+            ixx = pars[5]
+            iyy = pars[3]
+            ixy = pars[4]
```

Comment:
yes, I know this code is sloppy.  It needed to be rewritten, but not at the moment.  There is a general method which I need to write for generic ngmix models, but I was putting it off until DM-5432, since I don't have the information at the moment to write that method.  I will at least make this code more compact and clearer.